Add support for HKClinicalTypes

klandell opened the pull request:
## Description

Adds support for the HKClinicalTypes (https://developer.apple.com/documentation/healthkit/hkclinicaltype?language=objc)

This is an ***unfinished*** draft PR.  So far I have only fully implemented the getAllergyRecords function. Further work is needed, as well as documentation. If someone could look to see if this appears to be on the right track that would be amazing!

Fixes #69 

## Type of change

- [X] New feature (non-breaking change which adds functionality)
- [X] This change requires a documentation update

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] I have checked my code and corrected any misspellings


## Details
Requesting the following permission:

`AppleHealthKit.Constants.Permissions.AllergyRecord;`

and then calling

`AppleHealthKit.getAllergyRecords`

produces the following output with the Apple Health sample data
```JSON
[
  {
    "id": "54D63F66-FB1E-4DF3-852F-6D8C4648CE81",
    "displayName": "Peanuts",
    "fhirData": {
      "resourceType": "AllergyIntolerance",
      "recordedDate": "2015-02-18",
      "substance": {
        "text": "Peanuts",
        "coding": [
          {
            "code": "256349002",
            "system": "http://snomed.info/sct"
          }
        ]
      },
      "onset": "1995",
      "patient": {
        "display": "Candace Salinas",
        "reference": "Patient/1"
      },
      "reaction": [
        {
          "onset": "2015-02-17",
          "manifestation": [
            {
              "text": "Wheezing"
            }
          ],
          "severity": "severe"
        }
      ],
      "id": "2",
      "status": "active"
    },
    "endDate": "2021-07-01T15:11:09.041-0400",
    "startDate": "2021-07-01T15:11:09.041-0400",
    "sourceId": "com.apple.public.health.clinical.B0734C79-CD39-D049-84F6-CBE9A5629AB8",
    "sourceName": "Sample Location A"
  }
]
```

klandell commented:
Everything seems to be in place for the basic functionality.  Before converting to a real PR, I want to look into background processing for these new data types. I can get to that early next week.

klandell commented:
I still need to do some testing of this, but a code review at this point would be beneficial.

I added code for background processing. Though I haven't had any luck with that working for even the existing types.  Some of the clinical types are only available for specific os versions. Do I need to do anything special error handling? I didn't really see anything like that in the rest of the codebase,

GGGava commented:
Hey @klandell, thank you for your huge contribution. I'll try to review everything in the next few days.

Regarding the unavailable types, I faced a similar issue while working on the Swift migration. Check out how this is handled (if it is) in the activity types. Objective-C does not support reliable availability checks, so I'm not sure how the current code behaves in case of unavailability. You can handle the same way, but if you have another idea, let us know!

GGGava commented on RCTAppleHealthKit/RCTAppleHealthKit+Methods_ClinicalRecords.m:
```diff
@@ -0,0 +1,102 @@
+{
+    HKClinicalType *recordType = [RCTAppleHealthKit clinicalTypeFromName:type];
+    if (recordType == nil) {
+        callback(@[RCTMakeError(@"invalid type", nil, nil)]);
```
We could be more descriptive here

GGGava commented on RCTAppleHealthKit/RCTAppleHealthKit+Methods_ClinicalRecords.m:
```diff
@@ -0,0 +1,102 @@
+
+- (void)clinicalRecords_getAllergyRecords:(NSDictionary *)input callback:(RCTResponseSenderBlock)callback
+{
+    [self clinicalRecords_getClinicalRecordsOfType:@"AllergyRecord" input:input callback:callback];
```
what is the difference between calling these methods (lines 56-91) and calling `clinicalRecords_getClinicalRecords` with the type in the input?

klandell replied:
I definitely don't _disagree_ with you. In fact removing the specific getters (e.g. getAllergyRecords) will greatly reduce the size of this pull request because then I can delete a bunch of documentation :)

I added both the methods in (56-91) and `clinicalRecords_getClinicalRecords` to give the clinical records a similar API footprint to the fitness methods. At first glance I thought that calling for example `getSamples` with a type of `StepCount` was is equivalent to calling `getStepCount`, though on closer inspection they do return different results.

I'm definitely on board with simplifying the api to just the `getClinicalRecords` method

GGGava commented on RCTAppleHealthKit/RCTAppleHealthKit+Queries.m:
```diff
@@ -241,6 +241,57 @@ - (void)fetchSamplesOfType:(HKSampleType *)type
+- (void)fetchClinicalRecordsOfType:(HKClinicalType *)type
+                         predicate:(NSPredicate *)predicate
+                         ascending:(BOOL)asc
+                             limit:(NSUInteger)lim
```
We could change `asc` and `lim` to `ascending` and `limit` here, this is how these guys are named in other places, ok?

klandell replied:
It looks like all cases except `fetchWorkoutForPredicate` in this file use the convention of `ascending:(BOOL)asc` and `limit:(NSUInteger)lim` as used here.

GGGava replied:
No problems then 😄

GGGava reviewed (changes requested):
Great job so far!

Just left a few comments about some details, take a look :D

klandell commented:
> Great job so far!
> 
> Just left a few comments about some details, take a look :D

Thanks for taking a look!  I'll address your comments next week.

klandell commented:
@GGGava The commits since your review should have addressed all of your comments.  I haven't really had a chance to test out the background updates yet. It doesn't seem to work on simulator for any of the data types.  If I don't get to it by the end of the week I might just strip that part out.
